BPM: Use a separate thread pool

phw opened the pull request:
Use a separate thread pool for the BPM calculation.

Otherwise the metadata box would not update on selection changes, since it used the same thread pool for updating the UI.

Also call `QtCore.QCoreApplication.processEvents()` during the BPM calculation, this keeps the UI more responsive. Otherwise it gets really laggy.

Fixes #200

phw committed:
BPM: Use a separate thread pool

Otherwise the metadata box would not update on selection changes, since it used the same thread pool for updating the UI.

Fixes #200

phw commented:
@zas: What is your opinion on this? While this change works it is actually quite some work to implement.

I am thinking about introducing a `priority_thread_pool` in Picard instead, which would be used by operations where the user expects more or less instant feedback, in this case updating the metadata box when an item gets selected. Implementation of this inside Picard would be trivial (2 or 3 lines) and calculation heavy plugins like this one would not need to worry about it.

In case this comes up: I also tried to increase the priority of the tasks queued up by the metadata box update. This definitely improves the situation, but it is still very laggy. It could take over a second until the view updated, with a separate thread pool it was virtually instant.

zas commented:
> I am thinking about introducing a `priority_thread_pool` in Picard instead,

Yes, it might be better, limiting code redundancy if different plugins need the same feature.

madmouser1 reviewed (approved):
many thanks

phw commented:
@zas Looking at this at more detail I actually question that we need to start a thread at https://github.com/metabrainz/picard/blob/master/picard/ui/metadatabox.py#L449 at all.

The executed call is running Python code and due to the GIL does not run in parallel anyway. Maybe it would be better to call it without a thread and instead call `QtCore.QCoreApplication.processEvents()` inside the for loops in `_update_tags()` so UI events are executed even if comparing a lot of files.

phw commented:
Scratch that last idea, it behaves worse then the currently threaded solution. Just using a separate thread pool for it is a good idea.

But my testing showed also something different: For the BPM module it actually makes some sense to implement that thread pool handling itself, because it means it can also abort the threads cleanly (see the hanlding of `self._close`). Without that Picard waits for all threads to close. If you are calculating the BPM values for several hundred files this means the Picard process keeps running for quite a while after the window has already been closed.

So my current recommendation is merging this PR as is but in addition also commit the separate thread pool for UI updates to Picard.

zas commented:
> So my current recommendation is merging this PR as is but in addition also commit the separate thread pool for UI updates to Picard.

LGTM

phw commented:
Actually the calls to `QtCore.QCoreApplication.processEvents()` are nonsense here, since this only triggers processing of events in the current thread, and there are none. It would be an alternative to running in threads. I'll remove this call and merge then.

phw committed:
BPM: Do not use separate thread pool, but close threads on application close

This will work well with the updated implementation in Picard 2.1.3 while still giving the benefit of quick closing of the application.

phw commented:
@zas I updated this again and took a middle route. I did not implement a separate thread pool, but use the default one. This will work fine with the patch for Picard 2.1.3, but of course will require users to update Picard to benefit for it.

But I kept the code for early exit from the threads. This makes exiting Picard behave much nicer, since BPM analyzing can run for several minutes if you have it running for many files. Actually I think we should look at more places inside Picard where we could exit threads earlier.

What do you think?
